### DISCUSSION TITLE
multirotor fly in straight line

### DISCUSSION THREAD

**Discussion by Zzerr0r** [microsoft/AirSim#4497] [Support Q&A] [answered]
Hi, I am literally just trying to fly in a straight line.
I am using moveToPositionAsync(200,0,0,40) so i expected the drone to just fly 200 meters in x-direction and 0 in y- and z-direction. However the drone keeps dropping until it eventually crashes into the ground along its path.  I have also tried moveByVelocityAsync,  moveByVelocityZBodyFrameAsync and moveByVelocityBodyFrameAsync (whatever the difference between these 3 is --> is there no documentation or can i just not find it?) where i also set the z-velocity to 0, resulting in the drone crashing into the ground again (which is flat by the way). I tried changing the PhysicsEngineName setting in the settings.json to "ExternalPhysicsEngine" as suggested in another post (https://github.com/microsoft/AirSim/issues/2570), which resulted in the drone not moving at all anymore (as to be expected tbh). 
Any help is appreciated. Kinda losing my mind over something this basic. Thanks in advance!

**Comment by xxEoD2242** [comment, marked as answer]
> Hi, I am literally just trying to fly in a straight line.
> 
> I am using moveToPositionAsync(200,0,0,40) so i expected the drone to just fly 200 meters in x-direction and 0 in y- and z-direction. However the drone keeps dropping until it eventually crashes into the ground along its path.  I have also tried moveByVelocityAsync,  moveByVelocityZBodyFrameAsync and moveByVelocityBodyFrameAsync (whatever the difference between these 3 is --> is there no documentation or can i just not find it?) where i also set the z-velocity to 0, resulting in the drone crashing into the ground again (which is flat by the way). I tried changing the PhysicsEngineName setting in the settings.json to "ExternalPhysicsEngine" as suggested in another post (https://github.com/microsoft/AirSim/issues/2570), which resulted in the drone not moving at all anymore (as to be expected tbh). 
> 
> Any help is appreciated. Kinda losing my mind over something this basic. Thanks in advance!

Just give it -1 in the Z. You are telling it to crash into the ground with a 0 value for z in your moveByPositionAsync call

**Reply by Zzerr0r** [reply]
It is still crashing into the ground even with -1 in the Z

**Reply by xxEoD2242** [reply]
What's your starting Z value in the settings file?

Also, turn down the speed to say 5 m/s and increase z to -2.

**Reply by Zzerr0r** [reply]
Starting Z in settings file is 0, 
unfortunately my drone is required to fly alongside a car so I can't reduce its speed and puttingz to -2 by itself leads to the same result. even -20 won't do. I tried -100, which saves the drone from crashing but screws up the flight path

**Reply by xxEoD2242** [reply]
Well, you are telling the agent to go 90 mph, which is usually beyond the capabilities of a cascaded PID controller. We have had luck with 20 m/s before, which is what the speed input is in that moveByPosition command.

Beyond that, you would need your own controller that can handle that large of a delta between start and stop. Control theory tells us that the PID operates in a linear fashion and you are pushing the system (if you approximate behavior) into the nonlinear regime which makes it very difficult to control.

Model predictive control would be good to look at for cases such as these.

**Reply by Zzerr0r** [reply]
I see, guess I was using a wrong approach to my problem anyways. I have now switched to continuously teleporting the drone between taking pictures. Thanks for the explanation and your patience!

**Comment by xxEoD2242** [comment]
Well, 40 m/s is 90 miles per hour so I’m not sure what speed you are
attempting but my guess is the system is over correcting as it uses look
ahead. You need to look at how SimpleFlight works, as it is not a robust
flight control system. There are few systems I know of which could operate
a quad copter at 90 miles per hour.

To fix your issue, you would need to reduce your speed to somewhere around
20.

We have successfully operated a drone at 25 m/s before in AirSim but have
moved away from the Position Controller as we have found a number of
artifacts that are undesireable.
